tooling/minimize/src/main.rs

```diff
@@ -120,10 +129,147 @@ macro_rules! show_error {
     ($($tt:tt)*) => { crate::show_error(&format_args!($($tt)*)) };
 }
 
+pub fn get_sysroot_dir() -> PathBuf {
+    match std::env::var_os("MINIRUST_SYSROOT") {
+        Some(dir) => PathBuf::from(dir),
+        None => {
+            let user_dirs = directories::ProjectDirs::from("org", "rust-lang", "minirust").unwrap();
+            user_dirs.cache_dir().to_owned()
+        }
+    }
+}
+
+fn be_rustc() {
+    // Get the rest of the command line arguments
+    let mut args: Vec<OsString> = env::args_os().skip(1).collect();
+
+    let use_panic_abort = {
+        let mut yes = false;
+        let mut it = args.iter();
+        while let Some(a) = it.next() {
+            if let Some(s) = a.to_str() {
+                if s == "--crate-name" {
+                    if let Some(next) = it.next().and_then(|o| o.to_str()) {
+                        if next == "panic_abort" {
+                            yes = true;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        yes
+    };
+
+    // Inject the Rust flags
+    for arg in DEFAULT_ARGS {
+        args.push(arg.into());
+    }
+
+    // If we are building dependencies, inject the sysroot flag
+    if std::env::var_os("MINIMIZE_BUILD_DEPS").is_some() {
+        let sysroot_dir = setup_sysroot();
+        args.push(format!("--sysroot={}", sysroot_dir.display()).into());
+    }
+
+    args.push("-C".into());
+
+    if use_panic_abort {
+        args.push("panic=abort".into());
+    } else {
+        args.push("panic=unwind".into());
+    }
+
+    // Invoke the rust compiler
+    let status = Command::new("rustc")
+        .args(args)
+        .env_remove("RUSTC")
+        .env_remove("MINIMIZE_BE_RUSTC")
+        .env_remove("MINIMIZE_BUILD_DEPS")
+        .status()
+        .expect("failed to invoke rustc in custom sysroot build");
+
+    std::process::exit(status.code().unwrap_or(1));
+}
+
+fn setup_sysroot() -> PathBuf {
```

Comment:
Please move the sysroot setup logic into its own file.

Also, this function should have a doc comment explaining the meaning of the return value.